Makefile

```diff
@@ -71,65 +80,67 @@ orion-server-images:
 
+INTEGRATION_TARGETS = download-fabric integration-tests-iou
+INTEGRATION_TARGETS += integration-tests-atsacc 
+INTEGRATION_TARGETS += integration-tests-chaincode-events
+INTEGRATION_TARGETS += integration-tests-atsafsc
+INTEGRATION_TARGETS += integration-tests-twonets
+INTEGRATION_TARGETS += integration-tests-pingpong
+INTEGRATION_TARGETS += integration-tests-stoprestart
+
 .PHONY: integration-tests
-integration-tests:
-	cd ./integration/fabric/iou; ginkgo $(GINKGO_TEST_OPTS) .
-	cd ./integration/fabric/atsa/chaincode; ginkgo $(GINKGO_TEST_OPTS) .
-	cd ./integration/fabric/events/chaincode; ginkgo $(GINKGO_TEST_OPTS) .
-	cd ./integration/fabric/atsa/fsc; ginkgo $(GINKGO_TEST_OPTS) .
-	cd ./integration/fabric/twonets; ginkgo $(GINKGO_TEST_OPTS) .
-	cd ./integration/fsc/pingpong/; ginkgo $(GINKGO_TEST_OPTS) .
-	cd ./integration/fsc/stoprestart; ginkgo $(GINKGO_TEST_OPTS) .
+integration-tests: $(INTEGRATION_TARGETS)  
 
 .PHONY: integration-tests-iou
-integration-tests-iou:
-	cd ./integration/fabric/iou; ginkgo $(GINKGO_TEST_OPTS) .
+integration-tests-iou: download-fabric
+	cd ./integration/fabric/iou; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
 
 .PHONY: integration-tests-iou-hsm
-integration-tests-iou-hsm:
+integration-tests-iou-hsm: download-fabric
 	@echo "Setup SoftHSM"
 	@./ci/scripts/setup_softhsm.sh
 	@echo "Start Integration Test"
-	cd ./integration/fabric/iouhsm; ginkgo $(GINKGO_TEST_OPTS) .
+	cd ./integration/fabric/iouhsm; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
 
 .PHONY: integration-tests-iouorionbe
-integration-tests-iouorionbe: orion-server-images
-	cd ./integration/fabric/iouorionbe; ginkgo $(GINKGO_TEST_OPTS) .
+integration-tests-iouorionbe: orion-server-images download-fabric
+	cd ./integration/fabric/iouorionbe; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
 
 .PHONY: integration-tests-atsacc
-integration-tests-atsacc:
-	cd ./integration/fabric/atsa/chaincode; ginkgo $(GINKGO_TEST_OPTS) .
+integration-tests-atsacc: download-fabric
+	cd ./integration/fabric/atsa/chaincode; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
 
-integration-tests-chaincode-events:
-	cd ./integration/fabric/events/chaincode; ginkgo $(GINKGO_TEST_OPTS) .
+.PHONY: integration-tests-chaincode-events
+integration-tests-chaincode-events: download-fabric
+	cd ./integration/fabric/events/chaincode; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
 
 .PHONY: integration-tests-atsafsc
-integration-tests-atsafsc:
-	cd ./integration/fabric/atsa/fsc; ginkgo $(GINKGO_TEST_OPTS) .
+integration-tests-atsafsc: download-fabric
+	cd ./integration/fabric/atsa/fsc; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
 
 .PHONY: integration-tests-twonets
-integration-tests-twonets:
-	cd ./integration/fabric/twonets; ginkgo $(GINKGO_TEST_OPTS) .
+integration-tests-twonets: download-fabric
+	cd ./integration/fabric/twonets; export FAB_BINS=$(FAB_BINS); ginkgo $(GINKGO_TEST_OPTS) .
```

Comment:
If a developer runs the test locally, `FAB_BINS` is not required to be set manually, instead it triggers the `download-fabric` target and exports `FAB_BINS`. I am wondering what happens if the developer sets`FAB_BINS` and then runs `make integration-tests-X`. The `download-fabric` target is called and maybe something is downloaded (if not already exists) but then the `FAB_BINS` set by the developer is overwritten within the integration test target. Is it working like this or do I miss something?

Reply:
I would remove the `download-fabric` target as dep and just call it in our CI setup phase - similar what we do with our go tools.

Reply:
The only limitation here is using your own builds of fabric, otherwise if you want to try different versions of fabric then all you do is 
```
FABRIC_VERSION=2.4.6 make integration-tests
```
Without it being there, contributors to FSC have to perform the download-fabric themselves first (and manually do it if they want to change versions). I guess it depends on the usability aspects here of users here. However I think it's only for real experts that they would provide their own builds of fabric and as such I would think it would be trivial for them to run the tests and we could allow someone to override FAB_BINS as per your next suggestion

Happy to change it if you still prefer that

Reply:
Well, let's assume I have already somewhere my fab bins and I just want to use them. Apparently, there are users running on a platform that is not compatible with whatever you can download :)

If I now use `FAB_BINS` to override it, running a test target will work fine. However, the `download-fabric` target is still triggered and downloads the bins to the folder as specified in `FABRIC_BINARY_BASE`. Even if I don't need them, the download is triggered. I think that is an annoying behaviour.

I understand that we are adding another manual step to prep the system before running the tests. However, we already do that for the go-tools and docker images anyway. To make life easy, these targets could be bundled in a `setup-dev-env` (or call it similarly) target. 

Letting the user manually trigger the `download-fabric` target has also security implication. We don't want to force users to download "arbitrary" binaries and just execute them. By doing this step manually by default, users need to decide if they are fine with just downloading them or taking over the responsibility to get their own bins from somewhere. For CI this argument does not hold, but I kind of appreciate it if our build system does not download executables without actively asking me if this is what I want :)